File: js/pathfinder.js
import { GRID_H, GRID_W } from './config.js';
import { isWall, leftSpawnCells, rightGoalCells } from './map.js';

function walkable(col, row, blocked) {
  if (col < 0 || row < 0 || col >= GRID_W || row >= GRID_H) return false;
  if (isWall(col, row)) return false;
  if (blocked.has(`${col},${row}`)) return false;
  return true;
}

function neighbors(c, r) {
  return [[c, r + 1], [c, r - 1], [c + 1, r], [c - 1, r]];
}

function astar(start, goals, blocked) {
  const goalSet = new Set(goals.map(([c, r]) => `${c},${r}`));
  const startKey = `${start[0]},${start[1]}`;
  const open = new Map([[startKey, { c: start[0], r: start[1], g: 0 }]]);
  const cameFrom = new Map([[startKey, null]]);
  const gScore = new Map([[startKey, 0]]);

  const h = (c, r) => {
    let best = 1e9;
    for (const [gc, gr] of goals) {
      const d = Math.abs(c - gc) + Math.abs(r - gr);
      if (d < best) best = d;
    }
    return best;
  };

  while (open.size) {
    let curKey = null;
    let cur = null;
    let bestF = 1e9;
    for (const [key, node] of open) {
      const f = node.g + h(node.c, node.r);
      if (f < bestF) {
        bestF = f;
        curKey = key;
        cur = node;
      }
    }
    open.delete(curKey);

    if (goalSet.has(curKey)) {
      const path = [];
      let node = curKey;
      while (node != null) {
        const [nc, nr] = node.split(',').map(Number);
        path.push([nc, nr]);
        node = cameFrom.get(node);
      }
      return path.reverse();
    }

    for (const [nc, nr] of neighbors(cur.c, cur.r)) {
      if (!walkable(nc, nr, blocked)) continue;
      const nk = `${nc},${nr}`;
      const ng = cur.g + 1;
      if (ng < (gScore.get(nk) ?? 1e9)) {
        gScore.set(nk, ng);
        cameFrom.set(nk, curKey);
        open.set(nk, { c: nc, r: nr, g: ng });
      }
    }
  }
  return [];
}

export function findPath(blocked) {
  const starts = leftSpawnCells().filter(([c, r]) => walkable(c, r, blocked));
  const goals = rightGoalCells().filter(([c, r]) => walkable(c, r, blocked));
  if (!starts.length || !goals.length) return [];

  let best = [];
  for (const start of starts) {
    const path = astar(start, goals, blocked);
    if (path.length && (!best.length || path.length < best.length)) best = path;
  }
  return best;
}

export function findPathReverse(blocked) {
  const starts = rightGoalCells().filter(([c, r]) => walkable(c, r, blocked));
  const goals = leftSpawnCells().filter(([c, r]) => walkable(c, r, blocked));
  if (!starts.length || !goals.length) return [];

  let best = [];
  for (const start of starts) {
    const path = astar(start, goals, blocked);
    if (path.length && (!best.length || path.length < best.length)) best = path;
  }
  return best;
}

export function pathExistsWithTowerAt(col, row, towerCells) {
  const blocked = new Set(towerCells);
  blocked.add(`${col},${row}`);
  return findPath(blocked).length > 0;
}

File: game/pathfinder.py
"""A* pathfinding from left spawn zone to right goal zone."""

from __future__ import annotations

import heapq

from game.map_data import is_wall, left_spawn_cells, right_goal_cells


def _neighbors(c: int, r: int):
    for dc, dr in ((0, 1), (0, -1), (1, 0), (-1, 0)):
        yield c + dc, r + dr


def _walkable(col: int, row: int, blocked: set[tuple[int, int]]) -> bool:
    from game.config import GRID_H, GRID_W

    if not (0 <= col < GRID_W and 0 <= row < GRID_H):
        return False
    if is_wall(col, row):
        return False
    if (col, row) in blocked:
        return False
    return True


def find_path(blocked: set[tuple[int, int]]) -> list[tuple[int, int]]:
    """Shortest path from any left spawn cell to any right goal cell."""
    from game.config import GRID_H, GRID_W

    starts = [s for s in left_spawn_cells() if _walkable(s[0], s[1], blocked)]
    goals = {g for g in right_goal_cells() if _walkable(g[0], g[1], blocked)}
    if not starts or not goals:
        return []

    best_path: list[tuple[int, int]] = []
    for start in starts:
        path = _astar(start, goals, blocked, GRID_W, GRID_H)
        if path and (not best_path or len(path) < len(best_path)):
            best_path = path
    return best_path


def _astar(
    start: tuple[int, int],
    goals: set[tuple[int, int]],
    blocked: set[tuple[int, int]],
    gw: int,
    gh: int,
) -> list[tuple[int, int]]:
    def h(c, r):
        # distance to nearest goal column
        return min(abs(c - g[0]) + abs(r - g[1]) for g in goals)

    open_heap: list[tuple[int, int, tuple[int, int]]] = []
    heapq.heappush(open_heap, (h(*start), 0, start))
    came_from: dict[tuple[int, int], tuple[int, int] | None] = {start: None}
    g_score = {start: 0}

    while open_heap:
        _, g, current = heapq.heappop(open_heap)
        if current in goals:
            path = []
            node = current
            while node is not None:
                path.append(node)
                node = came_from[node]
            path.reverse()
            return path

        if g > g_score.get(current, float("inf")) + 200:
            continue

        for nc, nr in _neighbors(*current):
            if not _walkable(nc, nr, blocked):
                continue
            ng = g + 1
            if ng < g_score.get((nc, nr), float("inf")):
                g_score[(nc, nr)] = ng
                came_from[(nc, nr)] = current
                heapq.heappush(open_heap, (ng + h(nc, nr), ng, (nc, nr)))
    return []


def find_path_reverse(blocked: set[tuple[int, int]]) -> list[tuple[int, int]]:
    """Path from right goal zone to left spawn (for drones)."""
    from game.config import GRID_H, GRID_W

    starts = [s for s in right_goal_cells() if _walkable(s[0], s[1], blocked)]
    goals = {g for g in left_spawn_cells() if _walkable(g[0], g[1], blocked)}
    if not starts or not goals:
        return []

    best_path: list[tuple[int, int]] = []
    for start in starts:
        path = _astar(start, goals, blocked, GRID_W, GRID_H)
        if path and (not best_path or len(path) < len(best_path)):
            best_path = path
    return best_path


def path_exists_with_tower_at(
    col: int, row: int, tower_cells: set[tuple[int, int]]
) -> bool:
    blocked = set(tower_cells)
    blocked.add((col, row))
    return len(find_path(blocked)) > 0
